server/src/main/java/org/opensearch/common/cache/stats/StatsHolder.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.cache.stats;
+
+import org.opensearch.common.cache.ICacheKey;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.BiConsumer;
+
+/**
+ * A class caches use to internally keep track of their stats across multiple dimensions. Not intended to be exposed outside the cache.
+ */
+public class StatsHolder {
+
+    // The list of permitted dimensions. Should be ordered from "outermost" to "innermost", as you would like to
+    // aggregate them in an API response.
+    private final List<String> dimensionNames;
+
+    // A map from a set of cache stats dimension values -> stats for that ordered list of dimensions.
+    private final ConcurrentMap<Key, CacheStatsResponse> statsMap;
+
```

Comment:
How will `StatsHolder` find out about shards being removed? When an index is deleted, how do we clean up its cache stats?

Reply:
It should be possible to add some logic to drop all keys in StatsHolder containing a certain dimension value/name pair. On the IRC side this could be triggered within clear(CacheEntity cacheEntity). I will work on this over the weekend.